### PR TITLE
feat: allow bulk delete record that satisfy the filter conditon

### PR DESCRIPTION
**Pull request by KhaiCM**
# feat: allow bulk delete record that satisfy the filter conditon

### CODE REVIEW

File: resources.go [hunk 1]
```diff
@@ -106,6 +106,9 @@ func (a *App) createResources() error {
 		Add(fs.NewResource("delete", contentService.Delete, &fs.Meta{
 			Delete: "/:id",
 			Args:   fs.Args{"id": createArg(fs.TypeUint64, "The content ID")},
+		})).
+		Add(fs.NewResource("bulk-delete", contentService.BulkDelete, &fs.Meta{
+			Post: "/bulk-delete",
 		}))
```
Comment (ngocphuongnb) [inline resources.go:112]: We used the `DELETE /:id` for single record deletion. So I think we should use the `DELETE /delete` for the bulk deletion.
```suggestion
		Add(fs.NewResource("bulk-delete", contentService.BulkDelete, &fs.Meta{
			Delete: "/delete",
		}))
```

File: services/content/bulk_delete.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		return nil, errors.BadRequest(err.Error())
+	}
+
+	filter := c.Arg("filter")
+
+	predicates, err := db.CreatePredicatesFromFilterObject(
+		cs.DB().SchemaBuilder(),
+		model.Schema(),
+		filter,
+	)
```
Comment (ngocphuongnb) [inline services/content/bulk_delete.go:22]:
```suggestion
	predicates, err := db.CreatePredicatesFromFilterObject(
		cs.DB().SchemaBuilder(),
		model.Schema(),
		c.Arg("filter"),
	)
```

File: services/content/bulk_delete.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
+package contentservice
+
+import (
+	"github.com/fastschema/fastschema/db"
+	"github.com/fastschema/fastschema/fs"
+	"github.com/fastschema/fastschema/pkg/errors"
+)
+
+func (cs *ContentService) BulkDelete(c fs.Context, _ any) (any, error) {
```
Comment (ngocphuongnb) [inline services/content/bulk_delete.go:9]: This method should return the affected rows count.
```suggestion
func (cs *ContentService) BulkDelete(c fs.Context, _ any) (int, error) {
```

File: services/content/bulk_delete.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+	records, err := model.Query(predicates...).
+		Get(c.Context())
+
+	if err != nil {
+		return nil, errors.BadRequest(err.Error())
+	}
+
+	if len(records) == 0 {
+		return nil, errors.NotFound("no entities found")
```
Comment (ngocphuongnb) [inline services/content/bulk_delete.go:35]: If there is no records matched, this should stop without any error.
```suggestion
		return 0, nil
```

File: services/content/bulk_delete.go
Comment (ngocphuongnb) [inline services/content/bulk_delete.go:1]: Should we move the `BulkDelete` into `delete.go`?

File: services/content/bulk_delete.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		if err := isDeletable(int(recordID), schemaName); err != nil {
+			return nil, errors.InternalServerError("Cannot delete root user.")
+		}
+		ids = append(ids, int(recordID))
+	}
+
+	anyIDs := make([]any, len(ids))
+	for i, id := range ids {
+		anyIDs[i] = any(id)
+	}
```
Comment (ngocphuongnb) [inline services/content/bulk_delete.go:50]:
```suggestion
	var ids []any
	for _, record := range records {
		recordID := record.ID()
		if err := isDeletable(int(recordID), schemaName); err != nil {
			return nil, errors.InternalServerError("Cannot delete root user.")
		}
		ids = append(ids, any(recordID))
	}
```

File: services/content/bulk_delete.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+	if _, err := model.Mutation().Where(db.In("id", anyIDs)).Delete(c.Context()); err != nil {
```
Comment (ngocphuongnb) [inline services/content/bulk_delete.go:52]: I have concerns here. Should we delete by list of ids? Or use the filter condition in the delete query?

File: services/content/bulk_delete_test.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+func TestContentServiceBulkDeleteFail(t *testing.T) {
+	cs, server := createContentService(t)
+
+	blogModel := utils.Must(cs.DB().Model("blog"))
+	blogID := utils.Must(blogModel.CreateFromJSON(context.Background(), `{"name": "test blog"}`))
+	tagModel := utils.Must(cs.DB().Model("tag"))
+	tagJSON := fmt.Sprintf(`{"name": "test tag", "blogs_id": %d}`, blogID)
+	tagModel.CreateFromJSON(context.Background(), tagJSON)
```
Comment (ngocphuongnb) [inline services/content/bulk_delete_test.go:64]:
```suggestion
	_ = utils.Must(tagModel.CreateFromJSON(context.Background(), tagJSON))
```

**Comment by ngocphuongnb** [comment]
@KhaiCM Thanks for your contribution!
Please check my comments and let me know if there are any problems.

**Comment by KhaiCM** [comment]
@ngocphuongnb Thank you for your comment, I have refactored some of the code as you said. There is a comment from you related to this
```
Should we delete by list of ids? Or use the filter condition in the delete query?
```
I thought and found that because this is a function to delete based on filter conditions, it is important to get the id and check that id before deleting, so I kept the code as is. Please check again

File: services/content/content_test.go [hunk 8]
```diff
@@ -36,6 +36,46 @@ func createContentService(t *testing.T) (*cs.ContentService, *rr.Server) {
...
+				"type": "relation",
+				"name": "blogs",
+				"label": "Blogs",
+				"relation": {
+					"schema": "blog",
+					"field": "tags",
+					"type": "o2o",
+					"fk_columns": null,
+					"optional": false
+				}
```
Comment (ngocphuongnb) [inline services/content/content_test.go:78]: I think we should test with o2m relation, and the `tag.blogs` should be the relation owner.
```suggestion
			{
				"type": "relation",
				"name": "tags",
				"label": "Tags",
				"optional": true,
				"sortable": true,
				"relation": {
					"schema": "tag",
					"field": "blogs",
					"type": "o2m",
					"fk_columns": null,
					"optional": false
				}
			}
		]
	}`)
	utils.WriteFile(schemaDir+"/tag.json", `{
		"name": "tag",
		"namespace": "tags",
		"label_field": "name",
		"fields": [
			{
				"type": "string",
				"name": "name",
				"label": "Name",
				"sortable": true
			},
			{
				"type": "relation",
				"name": "blogs",
				"label": "Blogs",
				"relation": {
					"schema": "blog",
					"field": "tags",
					"type": "o2m",
					"owner": true,
					"fk_columns": null,
					"optional": false
				}
```

**Comment by khaicm-0897** [comment]
@ngocphuongnb I have revised the code according to your comment. In the data setup section, I have added *"on_delete": "restrict"* and the blog's relation to ensure that when deleting records with relations, there will be an error. Please help me check to see if this is okay!

**Comment by ngocphuongnb** [comment]
@KhaiCM Thank you!
Please remove the `on_delete` attribute from the test schema because it has no effect to fastschema relation.

File: services/content/content_test.go [hunk 9]
```diff
@@ -36,6 +36,47 @@ func createContentService(t *testing.T) (*cs.ContentService, *rr.Server) {
...
+				"label": "Tags",
+				"optional": true,
+				"sortable": true,
+				"relation": {
+					"schema": "tag",
+					"field": "blogs",
+					"type": "o2m",
+					"fk_columns": null,
+					"optional": false,
+					"on_delete": "restrict"
```
Comment (ngocphuongnb) [inline services/content/content_test.go:52]:
```suggestion
					"type": "o2m"
```

File: services/content/content_test.go [hunk 10]
```diff
@@ -36,6 +36,47 @@ func createContentService(t *testing.T) (*cs.ContentService, *rr.Server) {
...
+				"type": "relation",
+				"name": "blogs",
+				"label": "Blogs",
+				"relation": {
+					"schema": "blog",
+					"field": "tags",
+					"type": "o2m",
+					"owner": true,
+					"fk_columns": null,
+					"optional": false
```
Comment (ngocphuongnb) [inline services/content/content_test.go:78]:
```suggestion
					"owner": true
```

**Review by ngocphuongnb — APPROVED** [review]

**Comment by ngocphuongnb** [comment]
LGTM

**Pull request merged**